src/ucp/wireup/select.c

```diff
@@ -856,85 +866,51 @@ static double ucp_wireup_am_bw_score_func(ucp_context_h context,
     int is_proxy;
 
-    /* Check if we need active messages, for wireup */
-    if (!(ucp_ep_get_context_features(ep) & UCP_FEATURE_TAG) ||
-        (ep_init_flags & UCP_EP_INIT_FLAG_MEM_TYPE)) {
-        return UCS_OK;
-    }
-
-    /* Select one lane for active messages */
-    criteria.title              = "high-bw active messages";
-    criteria.local_md_flags     = 0;
-    criteria.remote_md_flags    = 0;
-    criteria.remote_iface_flags = UCT_IFACE_FLAG_AM_BCOPY |
-                                  UCT_IFACE_FLAG_CB_SYNC;
-    criteria.local_iface_flags  = UCT_IFACE_FLAG_AM_BCOPY;
-    criteria.calc_score         = ucp_wireup_am_bw_score_func;
-    criteria.tl_rsc_flags       = 0;
-    ucp_wireup_fill_ep_params_criteria(&criteria, params);
-
-    if (ucs_test_all_flags(ucp_ep_get_context_features(ep), UCP_FEATURE_TAG |
-                                                            UCP_FEATURE_WAKEUP)) {
-        criteria.local_iface_flags |= UCP_WORKER_UCT_UNSIG_EVENT_CAP_FLAGS;
-    }
-
     status             = UCS_ERR_UNREACHABLE;
-    num_lanes          = 1;
-    local_dev_bitmap   = -1;
-    remote_dev_bitmap  = -1;
-    md_map             = 0;
-
-    /* am_bw_lane[0] is am_lane, so don't re-select it here */
-    for (lane_desc_idx = 0; lane_desc_idx < *num_lanes_p; ++lane_desc_idx) {
-        if (lane_descs[lane_desc_idx].usage & UCP_WIREUP_LANE_USAGE_AM) {
-            addr_index         = lane_descs[lane_desc_idx].addr_index;
-            rsc_index          = lane_descs[lane_desc_idx].rsc_index;
-            md_map            |= UCS_BIT(context->tl_rscs[rsc_index].md_index);
-            local_dev_bitmap  &= ~UCS_BIT(context->tl_rscs[rsc_index].dev_index);
-            remote_dev_bitmap &= ~UCS_BIT(address_list[addr_index].dev_index);
-            break; /* do not continue searching due to we found
-                      AM lane (and there is only one lane) */
-        }
-    }
+    num_lanes          = 0;
+    md_map             = bw_info->md_map;
+    local_dev_bitmap   = bw_info->local_dev_bitmap;
+    remote_dev_bitmap  = bw_info->remote_dev_bitmap;
 
     /* lookup for requested number of lanes or limit of MD map
      * (we have to limit MD's number to avoid malloc in
      * memory registration) */
-    for (; (num_lanes < ep->worker->context->config.ext.max_eager_lanes) &&
+    for (; (num_lanes < bw_info->max_lanes) &&
            (ucs_count_one_bits(md_map) < UCP_MAX_OP_MDS);) {
         status = ucp_wireup_select_transport(ep, address_list, address_count,
-                                             &criteria, -1, -1,
+                                             &bw_info->criteria, -1, -1,
                                              local_dev_bitmap, remote_dev_bitmap,
                                              0, &rsc_index, &addr_index, &score);
         if (status != UCS_OK) {
             break;
         }
 
-        is_proxy = ucp_wireup_is_lane_proxy(ep, rsc_index,
+        is_proxy = allow_proxy &&
```

Comment:
we need proxy only for AM lanes, not for RMA lanes

Reply:
right, ```allow_proxy``` argument enables proxy for AM lanes only